beacon-chain/p2p/gossipsub_peer_crawler.go

```diff
@@ -0,0 +1,515 @@
+package p2p
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/OffchainLabs/prysm/v7/beacon-chain/p2p/gossipsubcrawler"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/semaphore"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+)
+
+type peerNode struct {
+	id       enode.ID
+	isPinged bool
+	node     *enode.Node
+	peerID   peer.ID
+	topics   map[gossipsubcrawler.Topic]struct{}
+}
+
+type crawledPeers struct {
+	g *GossipsubPeerCrawler
+
+	mu       sync.RWMutex
+	byEnode  map[enode.ID]*peerNode
+	byPeerId map[peer.ID]*peerNode
+	byTopic  map[gossipsubcrawler.Topic]map[peer.ID]struct{}
+}
+
+func (cp *crawledPeers) updateStatusToPinged(enodeID enode.ID) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	existingPNode, ok := cp.byEnode[enodeID]
+	if !ok {
+		return
+	}
+
+	// we only want to ping a node with a given NodeId once -> not on every sequence number change
+	// as ping is simply a test of a node being reachable and not fake
+	existingPNode.isPinged = true
+}
+
+func (cp *crawledPeers) removePeerOnPingFailure(enodeID enode.ID) {
+	cp.mu.Lock()
+	defer cp.mu.Unlock()
+
+	existingPNode, ok := cp.byEnode[enodeID]
+	if !ok {
+		return
+	}
+
+	// same idea as in "updateStatusToPinged" above.
+	// We don't want to test pings for every sequence number change for a given node as that
+	// can lead to an explosion in the number of pings the crawler needs to do.
+	// So, remove the peer when the first ping fails. If the node becomes reachable later,
+	// we will discover it during a re-crawl and ping it again to test for reachability.
+	// we're not blacklisting this peer anyways.
+	cp.updateTopicsUnlocked(existingPNode, nil)
+}
+
+func (cp *crawledPeers) updateCrawledIfNewer(node *enode.Node, topics []string) {
+	cp.mu.Lock()
```

Comment:
I'm honestly not a fan at all of all the `cp.mu.Unlock` spread all over the function.
It's actually SO easy to miss one and to have a deadlock.

IMO it's way better to create a sub function if needed starting with:

```golang
cp.mu.Lock()
defer cp.mu.Unlock()
```